third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1225,4 +1588,263 @@ void schedulePersistentKernel(Fusion* fusion, const ReductionParams& rparams) {
+  reduction_scheduler_utils::propagateTransformation(
+      inner_reference_tv, boundaryNodesSet);
+  reduction_scheduler_utils::propagateRFactor(
+      inner_reference_tv, inner_reduction_tvs[0], inner_reduction_tvs);
+  // For parallelization, we need to explicitly skip tvs in boundaryNodesSet,
+  // otherwise the parallelization will be propagated to the first domain of tvs
+  // in boundaryNodesSet, because that domain was calculated from the partial
+  // reduction and these tensors are not scheduled yet.
+  reduction_scheduler_utils::propagateParallelization(
+      fusion,
+      inner_reduction_tvs[0],
+      inner_reference_tv,
+      unroll,
+      vectorize,
+      is_outer_grid_persistence,
+      inner_reduction_tvs,
+      cached_inputs,
+      cached_outputs,
+      boundaryNodesSet);
+  // Propagate outer reduction. Each outer reduction is connected with its
+  // cached_gmem and output, since we added all the cached_gmem to the
+  // boundaryNodesSet, the transformation from one outer reduction can't
+  // propagate to other outer reductions due to the cutoff at boundaryNodesSet.
+  // Thus, we need a loop to initiate the propagation from each outer reduction.
+  // Parallelization will not propagate to inner reductions as they are
+  // transformed differently.
+  for (long unsigned int i = 0; i < outer_reference_tvs.size(); i++) {
+    reduction_scheduler_utils::propagateTransformation(
+        outer_reference_tvs[i], boundaryNodesSet);
+    reduction_scheduler_utils::propagateParallelization(
+        fusion,
+        outer_reduction_tvs[i],
+        outer_reference_tvs[i],
+        unroll,
+        vectorize,
+        is_outer_grid_persistence,
+        outer_reduction_tvs,
+        cached_inputs,
+        cached_outputs);
+  }
+
+  // special vectorization of temp gmem, vectorization_factor_tmp_gmem_write is
+  // guaranteed to be smaller or equal to input vectorization factor.
+  if (rparams.vectorization_factor_tmp_gmem_write > 1) {
+    for (auto tv : cached_gmem) {
+      TORCH_INTERNAL_ASSERT(
+          rparams.vectorization_factor_tmp_gmem_write <=
+              rparams.unroll_factor_inner_reduction,
+          "vectorization factor of temp gmem write should be smaller than that of inner reduction.")
+      if (rparams.vectorization_factor_tmp_gmem_write <
+          rparams.unroll_factor_inner_reduction) {
+        tv->split(-1, rparams.vectorization_factor_tmp_gmem_write);
+      }
+      tv->axis(-1)->parallelize(ParallelType::Vectorize);
+    }
+  }
+  // vectorization propagate through propagateParallelization only works for
+  // input and output tensors. propagate vectorization to cached_gmem_reload
+  // directly from output tv using parallelizeAllLike. must propagate seperaely
+  // for different tvs as outer reductions are transformed seperately.
+  if (rparams.vectorization_factor_outer > 1) {
+    for (auto tv : cached_gmem_reload) {
+      auto output_tvs = ir_utils::outputTvsOf(tv);
```

Comment:
Why is this guaranteed to work? Is the first output of `cached_gmem_reload` guaranteed to be a vectorized tensor? Shouldn't we find one from `cached_outputs`?

Reply:
guaranteed to work as the code is within `if (rparams.vectorization_factor_outer > 1)`. I added further protection:
with `findVectorizedOutputOf`

Reply:
My question is why the first output is the vectorized output. I'm not asking if the vectorization is done or not.

Reply:
runTimeCheck and scheduler ensures all the outputs are vectorized. If in the future we changed that part and allows unvectorized input/output, then `cached_gmem_reload` will also not be vectorized.

Reply:
This was due to my confusion with `outputTvsOf` and `consumerTvsOf`.